WordPress/openverse
# Calculate PR urgency based on ready for review time

## Fixes



Fixes #1984 by @sarayourfriend 

## Description


This turned out to be a good deal more complicated than I anticipated, primarily due to the need to find the date when previously drafted PRs were marked ready for the review. Accounting for this required creating whole new fixtures.

Along the way, I also noticed we had a small issue with our test fixtures that caused an incorrect test! We had the PR `number` and `id` set to the same thing. However, the `id` is GitHub's unique identifier for the PR across all of GitHub. It is used for some specific endpoints while the `number` is used for others. Changing the fixture to create IDs that are not the same as the PR "number" revealed a test that was accidentally using the "id" instead of the "number" to confirm a particular end state.

In testing this, I noticed another edge case that we do not handle in which PRs are skipped for reminders when they probably should not be: https://github.com/WordPress/openverse/issues/2026

That's a sufficiently complex issue that I didn't think it was wise to solve it in this already rather complicated PR.

## Testing Instructions

Check out the new unit tests and confirm they cover the different cases a PR could be in. The "artificial" test case is unfortunate in that the scenario it is testing isn't possible, but it's the easiest way to prove that PR urgency is deferred
to the "ready_for_review" event when it exists rather than the PR's creation date. Another way to test this is to call the `get_urgency_if_urgent` function directly but all the other tests are "integration" style rather than unit style and I didn't want to break that pattern if I could avoid it. However, if reviewers think it would make more sense to avoid creating an artificial scenario to prove the case, that is a viable alternative to testing the same concept.

Test this locally by setting `AIRFLOW_VAR_GITHUB_API_KEY` in you env to a github PAT and `AIRFLOW_VAR_PR_REVIEW_REMINDER_DRY_RUN=true`. The second one is critical if you want to avoid actually sending out pings!

When I ran it locally, I received these pings in the logs of the DAG:

```
[2023-05-04, 05:46:28 UTC] {pr_review_reminders.py:257} INFO - Pinging @zackkrida to review Add media type context to start of result link titles
[2023-05-04, 05:46:28 UTC] {pr_review_reminders.py:257} INFO - Pinging @obulat to review Project Proposal: Additional search views
[2023-05-04, 05:46:28 UTC] {pr_review_reminders.py:257} INFO - Pinging @dhruvkb to review Handle more than one value for a url query parameter
```

All three of these PRs, as of writing, are in the same boat. They were all created beyond the urgency window but have received updates within the window. They would not receive a ping under the current DAG but will correctly receive a ping under this new approach.

We don't have any real PRs at the moment that exercise the events list scenario, so hopefully the unit tests cover those cases sufficiently!

## Checklist



[best_practices]:
  https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin


<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>
